docs/proposals/20220414-runtime-hooks.md

```diff
@@ -0,0 +1,430 @@
+
+##### Example Request:
+
+```yaml
+apiVersion: hooks.runtime.cluster.x-k8s.io/v1alpha1
+kind: BeforeClusterDeleteRequest
+cluster:
+  apiVersion: cluster.x-k8s.io/v1beta1
+  kind: Cluster
+  metadata:
+   name: test-cluster
+   namespace: test-ns
+  spec:
+   ...
+  status:
+   ...
+```
+
+##### Example Response:
+
+```yaml
+apiVersion: hooks.runtime.cluster.x-k8s.io/v1alpha1
+kind: BeforeClusterDeleteResponse
+status: Success
+message: "error message if status == Failure"
+retryAfterSeconds: 10
+```
+
+For additional details, refer to the [Draft OpenAPI spec](https://editor.swagger.io/?url=https://gist.githubusercontent.com/ykakarap/6a58d976df90928cdb31768125d73613/raw/8ce6175060039c2fb7b3ea9a091d712228f56852/lifecycle-hooks-openapi-2.yaml).
+
+
+###  Runtime Extensions developer guide
+
+All guidelines defined in the [Runtime SDK](https://github.com/kubernetes-sigs/cluster-api/blob/b48a6ed07ac2bd353f99000270510369f4baa1a5/docs/proposals/20220221-runtime-SDK.md) apply to the implementation of Runtime Extensions of the hooks defined in this proposal. 
+
+TL;DR; Runtime Extensions are components that should be designed, written and deployed with great caution given that they can affect the proper functioning of the Cluster API runtime. A poorly implemented Runtime Extension could potentially block lifecycle transitions from happening.
+
+Following recommendations are especially relevant:
+
+* [Blocking and non Blocking](https://github.com/kubernetes-sigs/cluster-api/blob/b48a6ed07ac2bd353f99000270510369f4baa1a5/docs/proposals/20220221-runtime-SDK.md#blocking-hooks)
+* [Error management](https://github.com/kubernetes-sigs/cluster-api/blob/b48a6ed07ac2bd353f99000270510369f4baa1a5/docs/proposals/20220221-runtime-SDK.md#error-management)
+* [Avoid dependencies](https://github.com/kubernetes-sigs/cluster-api/blob/b48a6ed07ac2bd353f99000270510369f4baa1a5/docs/proposals/20220221-runtime-SDK.md#avoid-dependencies)
+
+
+###  Security Model
+
+For the general Runtime Extension security model please refer to the [developer guide in the Runtime SDK proposal](https://github.com/kubernetes-sigs/cluster-api/blob/75b39db545ae439f4f6203b5e07496d3b0a6aa75/docs/proposals/20220221-runtime-SDK.md#security-model).
+
+###  Risks and Mitigations
+
+####  Runtime Extension blocks Cluster lifecycle indefinitely
+
+Cluster lifecycle can be blocked indefinitely when a Runtime Extension either blocks or fails indefinitely. Mitigation:
+
+* Surface errors from the Runtime Extension that is blocking reconciliation in Conditions to drive necessary action needed by the user.
+* A Runtime Extension should be extensively unit and e2e tested to ensure it behaves as expected.
+* Users should be able to manually intervene and unblock the reconciliation.
+
+As future work, we will explore more options like circuit breaker and timeout to unblock reconciliation.
```

Comment:
How much extra scope would it be to add global (configurable) timeout enforcement to all runtime hooks? There are similar baked-in defaults to various k8s runtimes (20 minute context deadline wrappers, for example) that might be worth copying here to better ensure against forward progress deadlocks.

Reply:
I don't think the extra scope is the major issue here so much as user expectations - if users are setting RuntimeExtensions can we properly capture what users would expect from the system with a global timeout? 

It's something we're thinking about, but it's hard to design a good solution without having the system in place (at least minimally) and getting a good understanding for how it might be used and how it will interact with the other processes / reconciliation loops going on in Cluster API.

Reply:
Thanks for the thoughtful responses!